### PR TITLE
Re-organize command line options

### PR DESCRIPTION
**Pull request by mvz**
Work-in-progress implementation of #254.

Done:
* Verbosity boiled down to showing headings for fragrant files or not
* Line number options simplified
* Line number options are fully separated from `--ultra-verbose`: We can now have links showing, but line numbers hidden.

To do:
* [x] Is it sensible to have `-o` as short option for `--color`, or even have a short option at all, since it is the default?
* [x] Is `--ultra-verbose` a good name? How about `--show-links` or `--wiki-links`?
* [ ] Should `--single-line` be a fully separate kind of report
* [x] Perhaps we can replace all these format options with `--format`, with values `text`, `single-line`, `yaml`, `html`?
* [x] Replace `--sort-by-issue-count` with `--sort=..` option

### CODE REVIEW

**Comment by troessner** [comment]
>> Is it sensible to have -o as short option for --color, or even have a short option at all, since it is the default?

Excellent point, let's throw it out.

>> Is --ultra-verbose a good name? How about --show-links or --wiki-links?

I think I like "--wiki-links" better than "--ultra-verbose", let's go with that.:)

>> Should --single-line be a fully separate kind of report? Perhaps we can replace all these format options with --format, with values text, single-line, yaml, html?

Good idea! I also think that "--format" is the way to go.

**Comment by mvz** [comment]
I've sort of decided to keep `--single-line` as a suboption of `--format=text`.

**Comment by mvz** [comment]
I ended up doing a bit of refactoriong as well: There is now only one 'Command', since help and version are trivial to inline inside `Options`. There is more to be done (ReekCommand should absorb Command and be renamed; Options should be renamend).

I decided not to make `--single-line` into a separate format, since then it becomes very unclear which options apply to what.

And with that, I invite comments to this pull request.

**Comment by troessner** [comment]
I'll review this bad boy today..;)

File: features/command_line_interface/options.feature [hunk 1]
```diff
@@ -25,29 +25,33 @@ Feature: Reek can be controlled using command-line options
...
+          -U, --[no-]wiki-links            Show link to related Reek wiki page for each smell
+          -n, --[no-]line-numbers          Show line numbers in the output (this is the default)
+          -s, --single-line                Show location in editor-compatible single-line-per-smell format
+              --sort-by SORTING            Sort reported files by the given criterium:
+                                             smelliness ("smelliest" files first)
+                                             none (default - output in processing order)
+
+      Utility options:
+          -h, --help                       Show this message
+          -v, --version                    Show version
```
Comment (troessner) [inline features/command_line_interface/options.feature]: :+1:

File: features/reports/reports.feature [hunk 2]
```diff
@@ -51,6 +51,12 @@ Feature: Correctly formatted reports
       9 total warnings
       """
 
+    Examples:
+      | option         |
+      |                |
+      | --sort-by none |
+      | --sort-by n    |
```
Comment (troessner) [inline features/reports/reports.feature]: "--sort-by n" seems weird. Why not get rid of this?

Comment (troessner) [inline features/reports/reports.feature]: Actually why not get rid of the complete "none" option? What do we gain from it?

File: features/reports/reports.feature [hunk 3]
```diff
@@ -204,4 +228,20 @@ Feature: Correctly formatted reports
     Examples:
       | option           |
       | -U               |
-      | --ultra-verbose  |
+      | --wiki-links  |
+
+  Scenario Outline: --wiki-links is independent of --line-numbers
```
Comment (troessner) [inline features/reports/reports.feature]: A little confused here - why is this scenario in here? Seems arbitrary to test a random coupling of options.

File: lib/reek/cli/help_command.rb [hunk 4]
```diff
@@ -1,15 +0,0 @@
...
-    # A command to display usage information for this application.
-    #
-    class HelpCommand < Command
-      def execute(view)
-        view.output(@options.help_text)
-        view.report_success
-      end
-    end
-  end
-end
```
Comment (troessner) [inline lib/reek/cli/help_command.rb]: :1

File: lib/reek/cli/input.rb [hunk 5]
```diff
@@ -1,7 +1,7 @@
 require 'reek/source'
 
 module Reek
-  module CLI
+  module Cli
```
Comment (troessner) [inline lib/reek/cli/input.rb]: But "CLI" is an acronym, not a word. I have no hard feelings towards this but it seems a little strange to me...

File: lib/reek/cli/option_interpreter.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+      def report_class
+        case @options.report_format
+        when :yaml
+          Report::YamlReport
+        when :html
+          Report::HtmlReport
+        else # :text
+          Report::TextReport
+        end
```
Comment (troessner) [inline lib/reek/cli/option_interpreter.rb]: Very nice!

File: lib/reek/cli/option_interpreter.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      end
+
+      def warning_formatter
+        klass = if @options.show_links
+                  Report::UltraVerboseWarningFormatter
```
Comment (troessner) [inline lib/reek/cli/option_interpreter.rb]: Then we probably should rename the Formatter as well to "ShowLinksWarningFormatter" - what do you think?

File: lib/reek/cli/option_interpreter.rb [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        end
+      end
+
+      def sort_by_issue_count
+        case @options.sorting
+        when :smelliness
+          true
+        else
+          false
+        end
```
Comment (troessner) [inline lib/reek/cli/option_interpreter.rb]: Why not just 

```Ruby
def sort_by_issue_count
  @options.sorting == :smelliness
end
```

File: lib/reek/cli/options.rb [hunk 9]
```diff
@@ -14,110 +9,122 @@ module Cli
...
+        @options.argv = @argv
+        Rainbow.enabled = @options.colored
+        @options
       end
 
+      private
+
       def banner
-        progname = @parser.program_name
+        program_name = @parser.program_name
```
Comment (troessner) [inline lib/reek/cli/options.rb]: Ah, special thanks for that.:)

File: lib/reek/cli/options.rb [hunk 10]
```diff
@@ -14,110 +9,122 @@ module Cli
...
+        set_configuration_options
+        set_alternative_formatter_options
+        set_report_formatting_options
+        set_utility_options
+      end
+
+      def set_utility_options
+        @parser.separator "\nUtility options:"
+        @parser.on_tail('-h', '--help', 'Show this message') do
+          puts @parser
```
Comment (troessner) [inline lib/reek/cli/options.rb]: I like this a lot. Using separate commands for this always seemed weird.

File: lib/reek/cli/options.rb [hunk 11]
```diff
@@ -14,110 +9,122 @@ module Cli
...
-        @parser.separator 'Common options:'
-        @parser.on('-h', '--help', 'Show this message') do
-          @command_class = HelpCommand
+        set_configuration_options
+        set_alternative_formatter_options
+        set_report_formatting_options
+        set_utility_options
+      end
+
+      def set_utility_options
```
Comment (troessner) [inline lib/reek/cli/options.rb]: Can we order the methods below the way they are used in set_up_parser? Just makes it a little easier to follow the flow.

**Comment by troessner** [comment]
Reviewed, looks great overall.

File: features/reports/reports.feature (hunk 2 above)
Comment (mvz) [inline features/reports/reports.feature]: At some point we may want to support default CLI arguments in the reek configuration (I think there's an issue for that). Then, it becomes necessary to be able to undo the `--sort-by smelliness`. The `n` bit is to show that OptionParser will accept unique prefixes for these parameters and these will be handled correctly. That bit is not really an important feature so I can take it out of the scenario.

File: lib/reek/cli/input.rb (hunk 5 above)
Comment (mvz) [inline lib/reek/cli/input.rb]: It's `Cli` everywhere else in the codebase. I can change it all to `CLI` if you prefer. I agree that would make more sense.

File: lib/reek/cli/option_interpreter.rb (hunk 8 above)
Comment (mvz) [inline lib/reek/cli/option_interpreter.rb]: D'oh! I'll change it.

File: lib/reek/cli/option_interpreter.rb (hunk 7 above)
Comment (mvz) [inline lib/reek/cli/option_interpreter.rb]: I agree.

File: features/reports/reports.feature (hunk 3 above)
Comment (mvz) [inline features/reports/reports.feature]: In 1.6, you cannot have wiki links with no line numbers or with the single-line format.

File: features/reports/reports.feature (hunk 2 above)
Comment (troessner) [inline features/reports/reports.feature]: Ok, understood - then let's leave it like it is.

File: features/reports/reports.feature (hunk 3 above)
Comment (troessner) [inline features/reports/reports.feature]: Ah..:)

File: lib/reek/cli/input.rb (hunk 5 above)
Comment (troessner) [inline lib/reek/cli/input.rb]: Alright, then let's leave it like it is and fix the spelling in a separate PR some time later.

**Comment by troessner** [comment]
Looking good! Please squash so I can merge.

**Comment by mvz** [comment]
Done!

**Comment by guilhermesimoes** [comment]
Don't forget to update the README regarding the format options when you release a new version.

**Comment by troessner** [comment]
Great catch @guilhermesimoes ! I just created https://github.com/troessner/reek/issues/383 so we don't forget to update this.

**Pull request merged**